1-Draft/RFCNNNN-Dispose-Script-Block.md

```diff
@@ -0,0 +1,279 @@
+    - This includes being executed after `StopProcessing()` is called, and not being interrupted/cancelled by `StopProcessing()` calls; if the command is being disposed, it has already completed processing.
+
+### Behaviour
+
+The `cleanup {}` block silently **discards** anything submitted to the success output stream from within it.
+
+Documentation should advise that `cleanup {}` is _not intended_ for any standard output and will completely ignore it.
+Additionally, all other PowerShell data streams will be accessible from `cleanup {}`, meaning there is no issue sending data using any of the following streams from within it:
+
+- Error
+- Warning
+- Verbose
+- Debug
+- Information
+
+### Implementation Details
+
+- A new `cleanup` keyword, indicating a `NamedBlockAst` similarly to `begin`, `process`, and `end`, which is valid in **all the same contexts** as the existing named block keywords.
+  - This includes new token types and appropriate entries in the `Tokenizer`'s keyword tables.
+- Additional members and constructors for `ScriptBlockAst` to enable it to be recognised by the parser.
+- Additional methods in `DlrScriptCommandProcessor` and `CompiledScriptBlock` classes to handle the block execution.
+  - Both types implement or inherit from a type that implements `IDisposable`, and as such the existing `Dispose()` methods can generally be utilised to run this block.
+- Implementing `IDisposable` for `PSScriptCmdlet` to handle the invocation for the `cleanup {}` block.
+- Changes to the behaviour of command execution in `CommandProcessorBase` to recognise when a function or script cmdlet is being disposed, and dispose of these appropriately via a new `DisposeScriptCommands()` method.
+  - Also required with this is an additional layer of `try/finally` during `Dispose()` events as we are opening up the possibility for users to throw terminating errors _during_ a command's disposal, and as such we must ensure that the engine-side disposal routines still execute appropriately.
+- `PipelineProcessor` will `Dispose()` commands immediately after their `EndProcessing()` segments complete to ensure resources are disposed as soon as they're no longer needed.
+- Changes will be required for `.ForEach{}` and `.Where{}` magic method implementations, which have some custom handling for named blocks that will need to be updated to account for `cleanup {}`.
+- Improvements to `PipelineStopper` to be able to properly recognise critical regions of code (like `cleanup {}`) and prevent pipeline stops from being registered until that region of code is exited.
+  - Ctrl+C will still be registered, but the pipeline stop will not begin until that critical code region is exited.
+
+#### Error States
+
+Terminating errors can be thrown from a `cleanup {}` block, which will skip the invocation of the rest of that `cleanup {}` block, without preventing other commands' `cleanup {}` blocks from being executed.
```

Comment:
Quick question related to this: what happens with the following statements in a cleanup block:

- `throw "bad"`
- `$PSCmdlet.ThrowTerminatingError(...)`
- `$PSCmdlet.WriteError(...)`

Which automatic variables will be available, and will any of them have fields that don't make sense or otherwise don't work in the cleanup block?

Reply:
Non-terminating errors work normally. Terminating errors work OK, but only if the command hasn't already thrown an error out.

In terms of the current implementation that I have in the code PR itself, I've noticed that there's an issue with reporting terminating errors in the Cleanup{} block due to the design of the pipeline processor _if_ (and only if) the command has already thrown a terminating error. Basically, the error is recorded in `$error` and then the pipeline processor suppresses it, because the command is already stopping due to an earlier error. Same should happen with errors thrown from `Dispose()` in compiled cmdlets, I think, even in the original code there.

I'd rather avoid refactoring that out of the pipeline processor itself, it seems like that would be _super_ messy. If there's a way I can catch that coming out of the `cleanup {}` block (or indeed the `Dispose()` block from compiled commands) and make it a non-terminating error at _that_ point (so the `cleanup{}` is still cancelled by a terminating error, but the pipeline reframes the error as non-terminating if it's already received a terminating error from that command, and the command has already finished what it can do anyway).

Something like that, anyway. Hopefully that makes sense. 😁

Reply:
Yeah that all makes sense. At the very least the pipeline processor is so central that any issues that come about when making the necessary modifications should present themselves pretty quickly.

With the variables I'm particularly thinking of things like `$PSCmdlet` and `$ExecutionContext` that have deep properties hooking back into the engine -- I expect that at least some of them will behave quite differently in a cleanup block and we should do our best to establish *how*, or if they will have parts that shouldn't be relied upon (or should behave differently -- e.g. throw because it's not a valid time to use them).

Reply:
In order for the cleanup block to run at all, we essentially have to tell the pipeline stopper to temporarily pretend all's fine and dandy until we're done with the cleanup block.

As a result, all those variables seem to behave themselves just fine, at least at the moment. But yeah, depending on decisions made during code review, that might change a bit, we'll see how it goes, definitely something to keep a sharp eye on. 🙂